src/components/z-carousel/index.tsx

```diff
@@ -1,36 +1,284 @@
+    }
+  }
+
+  /**
+   * Set an intersection observer to show the current index to the indicator when scrolling.
+   */
+  private setIntersectionObserver(): void {
+    this.intersectionObserver = new window.IntersectionObserver(
+      (entries) => {
+        const entry = entries.find((entry) => entry.isIntersecting);
+        if (!entry) {
+          return;
+        }
+
+        const entryIndex = this.items.findIndex((item) => item === entry.target);
+
+        /* skip setting the current item if intersection has been triggered by a programmatic scroll
+        (the scrollIntoView in `onIndexChange`) */
+        if (this.scrollingTo !== null && entryIndex !== this.scrollingTo) {
+          return;
+        }
+
+        this.current = entryIndex;
+        this.scrollingTo = null;
+      },
+      {
+        root: this.itemsContainer,
+        threshold: 0.5,
+      }
+    );
+
+    this.items.forEach((element) => this.intersectionObserver.observe(element));
+  }
+
+  private onPrev(): void {
+    if (this.single) {
+      this.current = Math.max(0, this.current - 1);
+      return;
+    }
+
+    const scroller = this.itemsContainer;
+    if (!scroller) {
+      return;
+    }
+
+    scroller.scrollBy({
+      left: -scroller.clientWidth / 2,
+      behavior: "smooth",
+    });
+  }
+
+  private onNext(): void {
+    if (this.single) {
+      this.current = Math.min(this.current + 1, this.items.length - 1);
+
+      return;
+    }
+
+    const scroller = this.itemsContainer;
+    if (!scroller) {
+      return;
+    }
+
+    scroller.scrollBy({
+      left: scroller.clientWidth / 2,
+      behavior: "smooth",
+    });
+  }
+
+  /**
+   * Check if footer can be rendered.
+   * @returns {boolean}
+   */
+  private canShowFooter(): boolean {
+    return (
+      this.arrows === CarouselArrowsPosition.BOTTOM ||
+      this.progress === CarouselProgressMode.DOTS ||
+      this.progress === CarouselProgressMode.NUMBERS
+    );
+  }
+
+  /**
+   * Set current item to passed index.
+   * @param {number} index Index to set
+   * @returns {void}
+   */
+  private goTo(index): void {
+    if (this.current === index) {
+      return;
+    }
+
+    this.current = index;
+  }
+
+  render(): HTMLDivElement | HostElement {
     if (this.isLoading) {
-      return (
+      return <Host>
+        {this.label && <div class="heading-4 z-carousel-title">{this.label}</div>}
         <div style={{ height: `${this.ghostLoadingHeight}px` }}>
           <z-ghost-loading></z-ghost-loading>
-          <div style={{ display: "none" }}>
+          <div class="loading-items-container">
             <slot />
           </div>
         </div>
-      );
+      </Host>;
     }
+
     return (
-      <ul class="z-carousel-items-container">
-        <slot />
-      </ul>
+      <Host>
+        <div class="z-carousel-container">
+          {this.label && <div class="heading-4 z-carousel-title">{this.label}</div>}
+          {this.arrows === CarouselArrowsPosition.OVER && (
```

Comment:
Invece di valutare la condizione due volte, si potrebbe valutarla una volta sola salvandola in una costante da riutilizzare?